### ISSUE TITLE
Support for new Amazon ElasticCache Auto Discovery Feature

### ISSUE THREAD

**Issue by thrillcall** [petergoldstein/dalli#296]
Would it be possible to add support for a new feature of the AWS ElastiCache (Memcache) service to auto-discover hosts in a memcached cluster?

http://aws.typepad.com/aws/2012/11/amazon-elasticache-now-with-auto-discovery.html
# Description:

Amazon ElastiCache supports Auto Discovery—the ability for client programs to automatically identify all of the nodes in a cache cluster, and to initiate and maintain connections to all of these nodes. With Auto Discovery, your application does not need to manually connect to individual cache nodes; instead, your application connects to a config endpoint. The config endpoint DNS entry contains the CNAME entries for each of the cache node endpoints; thus, by connecting to the config endpoint, you application immediately "knows" about all of the nodes in the cluster and can connect to all of them. You do not need to hardcode the individual cache node endpoints in your application.

All of the cache nodes in the cluster maintain a list of metadata about all of the other nodes. This metadata is updated whenever nodes are added or removed from the cluster.

Auto Discovery offers the following benefits:

When you scale up a cache cluster, the new nodes register themselves with the config endpoint and with all of the other nodes. When you scale down the cache cluster, the departing nodes deregister themselves. In both cases, all of the other nodes in the cluster are updated with the latest cache node metadata.

Cache node failures are automatically detected; failed nodes are automatically marked as unavailable.

A client program only needs to connect to the config endpoint. After that, the Auto Discovery library connects asynchronously to all of the other nodes in the cache cluster.

Client programs poll the cluster once per minute (this interval can be adjusted if necessary). If there are any changes to the cluster configuration, such as new or deleted nodes, the client receives an updated list of metadata. The client then connects to, or disconnects from, these nodes as needed.
# Supporting Documentation:

http://docs.amazonwebservices.com/AmazonElastiCache/latest/UserGuide/AutoDiscovery.HowAutoDiscoveryWorks.html

http://docs.amazonwebservices.com/AmazonElastiCache/latest/UserGuide/AutoDiscovery.ConfigCommand.html

Cheers,

Glenn

**Comment by kwent** [comment]
Would be nice to see this feature supported out of the box. Ideally the `dallu` client would refresh the nodes with a specified interval in a thread safe way.

**Comment by waissbluth** [comment]
are there any plans to incorporate this?

**Comment by pda** [comment]
From [a decade upthread](https://github.com/petergoldstein/dalli/issues/296#issuecomment-12396378) in 2013:

> > To support Auto Discovery, Amazon ElastiCache has added a new command to the Memcached ASCII protocol: config.
>
> Dalli only supports the binary protocol.

FYI, sometime since then ElastiCache Auto Discovery has been added to the memcached binary protocol:

> For cache engine version 1.4.14 or higher, use the config command. This command has been added to the Memcached ASCII **and binary** protocols by ElastiCache, and is implemented in the ElastiCache Cluster Client. If you want to use Auto Discovery with another client library, then that library will need to be extended to support the config command.
— https://docs.aws.amazon.com/AmazonElastiCache/latest/mem-ug/AutoDiscovery.AddingToYourClientLibrary.html#AutoDiscovery.AddingToYourClientLibrary.1-4-14-plus

**Comment by petergoldstein** [comment]
Hi @pda . I don't really understand the point of the comment.  As noted above, this capability is already available using [`dalli-elasticache`](https://github.com/ktheory/dalli-elasticache) .  There's no intention (or need) to add this to `dalli`.

The issue with incorporating is not the protocol type, but rather the functionality.  `Dalli::Client` gets a fixed set of memcached endpoints upon initialization, and that set is static for the lifetime of the object.  Individual endpoints in the set may become unreachable during the lifetime of the `Dalli::Client` object, but there's no current ability to add to that set (or explicitly remove for that matter).

So the way to incorporate Auto Discovery into a workflow is to use `dalli-elasticache` to generate `Dalli::Client` objects, and to occasionally (frequency based on use case and frequency of changes to the memcached cluster) replace long-lived `Dalli::Client` objects with new ones.

**Comment by pda** [comment]
> I don't really understand the point of the comment.

For background, I recently experienced an issue where a Dalli client was accidentally pointed at an ElastiCache configuration endpoint, rather 1–N individual nodes, and was randomly switching nodes based on DNS. That led me to search Dalli open+closed issues for ElastiCache Auto Discovery support, which landed on this issue describing just that.

This issue was closed in 2013 because “Auto Discovery uses ASCII protocol … Dalli only supports the binary protocol”, which I took to be the reason for not further exploring adding Auto Discovery directly in `dalli`, and the reason `dalli-elasticache` came to be.

Nobody had since mentioned that the Auto Discovery functionality is now available in the binary protocol too, so it seemed worth adding to this thread as an FYI.

But your rationale for preferring ElastiCache Auto Discovery support remain in `dalli-elasticache` regardless of the initially mentioned ASCII/binary protocol concerns makes total sense, and I'm glad it's all described in this thread now, thanks! 😁
